add stride constraints to fallbacks

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88706
* __->__ #88534

Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
add stride constraints to fallbacks

[ghstack-poisoned]

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



[ghstack-poisoned]

ngimel commented on torch/_inductor/ir.py:
```diff
@@ -104,6 +108,25 @@ def reads_from_conv(buf, var_ranges):
     return False, None
 
 
+def ir_node_to_tensor(x, guard_shape=True):
```
so for now we still expect fallback kernels to hardcode the shapes? We'd need to address this for dynamic shapes

eellison replied:
That's what the current state is, yea. But I imagine the change set to enable dynamic shapes here will be pretty small - construct the FakeTensors with the symbolic shapes instead of the real ones

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -f "Flakey windows test"

eellison committed:
Update on "add stride constraints to fallbacks"


Add stride/contiguity constraints to fallbacks so that inputs will be in the right stride permutation for the fallback kernel. 

Improves perf of coat_lite_mini from 1.48415536054865 -> 2.010956856330101. 



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel commented on torch/_inductor/ir.py:
```diff
@@ -2373,33 +2402,39 @@ def realize_input(cls, x):
+                x.get_layout(), FlexibleLayout
+            ) and is_stride_order_storage_and_layout(x, order):
+                # fix flexiblelayout to be FixedLayout with stride_order
+                as_storage_and_layout(
```
how does this work if you don't get return value of `as_storage_and_layout`?

eellison replied:
`freeze=True` inplace modifies it

eellison commented:
@pytorchbot merge

XiaobingSuper commented:
@eellison , I find that this PR introduces a redundant memory copy, see the following test case:

```
import torch
import torchvision
import torch._dynamo
import torch.fx.experimental.optimization as optimization
import copy
import time
import torch.profiler as profiler
from torch.fx import symbolic_trace
from torch._inductor import config
config.debug = True
from torch.utils import mkldnn as mkldnn_utils
#torch.manual_seed(2020)
torch._dynamo.config.verbose=True
#torch._dynamo.config.suppress_errors = True

class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv = torchvision.models.resnet50().conv1

    def forward(self, x):
        x = self.conv(x)
        return x

model= Model().eval()

model = model.to(memory_format=torch.channels_last).eval()
print(model)
with torch.no_grad():
    opt_model = torch._dynamo.optimize('inductor')(model)


batch_size = 20
x = torch.randn(batch_size, 3, 224, 224)
x = x.to(memory_format=torch.channels_last)

with torch.no_grad():
    for i in range(10):
        y1 = opt_model(x)

```

before this PR:

```
from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, arg1_1 = args
    args.clear()
    buf0 = aten.convolution(arg1_1, arg0_1, None, (2, 2), (3, 3), (1, 1), False, (0, 0), 1)
    assert_size_stride(buf0, (20, 64, 112, 112), (802816, 1, 7168, 64))
    del arg0_1
    del arg1_1
    return (buf0, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((64, 3, 7, 7), (147, 1, 21, 3), device='cpu', dtype=torch.float32)
    arg1_1 = rand_strided((20, 3, 224, 224), (150528, 1, 672, 3), device='cpu', dtype=torch.float32)
    print_performance(lambda: call([arg0_1, arg1_1]))
```

this PR:

```
from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()


kernel0 = async_compile.cpp('''
#include "/tmp/torchinductor_xiaobing/ik/cikrybpw4xhois4wll6h5afsswjrhpsb6gslcxrntzqtlyw2btey.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(40)
    {
        #pragma omp for
        for(long i0=0; i0<3010560; i0+=1)
        {
            {
                {
                    auto tmp0 = in_ptr0[i0];
                    out_ptr0[i0] = tmp0;
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, arg1_1 = args
    args.clear()
    buf1 = empty_strided((20, 3, 224, 224), (150528, 1, 672, 3), device='cpu', dtype=torch.float32)
    kernel0(c_void_p(arg1_1.data_ptr()), c_void_p(buf1.data_ptr()))
    del arg1_1
    buf0 = aten.convolution(buf1, arg0_1, None, (2, 2), (3, 3), (1, 1), False, (0, 0), 1)
    assert_size_stride(buf0, (20, 64, 112, 112), (802816, 1, 7168, 64))
    del arg0_1
    return (buf0, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((64, 3, 7, 7), (147, 1, 21, 3), device='cpu', dtype=torch.float32)
    arg1_1 = rand_strided((20, 3, 224, 224), (150528, 1, 672, 3), device='cpu', dtype=torch.float32)
    print_performance(lambda: call([arg0_1, arg1_1]))
```

XiaobingSuper commented on torch/_inductor/ir.py:
```diff
@@ -2373,33 +2402,39 @@ def realize_input(cls, x):
-            order
-        ):
-            return x
+        if is_storage_and_layout(x):
```
**is_storage_and_layout** need x is TensorBox, StorageBox or ReinterpretView, so if x is **InputBuffer** or **ComputedBuffer**, it always has a copy, I think it doesn't make sense.

eellison replied:
Good catch. I think the reason at the moment is that InputBuffer doesn't have a storage, but there's no reason it shouldn't (see: [TODO: input aliasing](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/graph.py#L202).) 

However this particular problem (copy on the input) will get fixed by making the output here a FixedLayout in https://github.com/pytorch/pytorch/pull/88706 because we no longer call `apply_constraints`.
